File: src/inverter/switch.py
from collections import namedtuple
from pathlib import Path

# GPIO sysfs paths
GPIO_EXPORT = Path("/sys/class/gpio/export")
GPIO_UNEXPORT = Path("/sys/class/gpio/unexport")
GPIO_BASE = Path("/sys/class/gpio")

# Define switch state structure
SwitchState = namedtuple(
    "SwitchState",
    ["pin", "is_closed"],
)

# Default GPIO pins for the 4 switch wires (adjust for your BBB setup)
DEFAULT_PINS = [66, 67, 68, 69]  # P8_7, P8_8, P8_9, P8_10


def export_pin(pin):
    """Export a GPIO pin for userspace control."""
    gpio_path = GPIO_BASE / f"gpio{pin}"
    if not gpio_path.exists():
        GPIO_EXPORT.write_text(str(pin))


def unexport_pin(pin):
    """Unexport a GPIO pin."""
    gpio_path = GPIO_BASE / f"gpio{pin}"
    if gpio_path.exists():
        GPIO_UNEXPORT.write_text(str(pin))


def set_direction(pin, direction="out"):
    """Set GPIO pin direction ('in' or 'out')."""
    direction_path = GPIO_BASE / f"gpio{pin}" / "direction"
    direction_path.write_text(direction)


def set_value(pin, value):
    """Set GPIO pin value (0 or 1)."""
    value_path = GPIO_BASE / f"gpio{pin}" / "value"
    value_path.write_text(str(value))


def get_value(pin):
    """Get GPIO pin value."""
    value_path = GPIO_BASE / f"gpio{pin}" / "value"
    return int(value_path.read_text().strip())


def init_switch(pin):
    """Initialize a single switch pin."""
    export_pin(pin)
    set_direction(pin, "out")
    set_value(pin, 0)  # Start open
    return SwitchState(pin=pin, is_closed=False)


def init_switches(pins=None):
    """Initialize all switch pins. Returns list of SwitchState."""
    if pins is None:
        pins = DEFAULT_PINS
    return [init_switch(pin) for pin in pins]


def close_switch(switch):
    """Close the switch (short the wires)."""
    set_value(switch.pin, 1)
    return SwitchState(pin=switch.pin, is_closed=True)


def open_switch(switch):
    """Open the switch (disconnect the wires)."""
    set_value(switch.pin, 0)
    return SwitchState(pin=switch.pin, is_closed=False)


def set_switch(switch, closed):
    """Set switch state. Returns new SwitchState."""
    if closed:
        return close_switch(switch)
    else:
        return open_switch(switch)


def cleanup_switches(switches):
    """Cleanup and unexport all switch pins."""
    for switch in switches:
        set_value(switch.pin, 0)
        unexport_pin(switch.pin)


def print_switches(switches):
    """Print the state of all switches."""
    for i, switch in enumerate(switches):
        state = "CLOSED" if switch.is_closed else "OPEN"
        print(f"Switch {i} (GPIO {switch.pin}): {state}")

File: src/inverter/hs/src/Switch.hs
module Switch (
    SwitchState (..),
    Pin,
    defaultPins,
    initSwitch,
    initSwitches,
    closeSwitch,
    openSwitch,
    setSwitch,
    cleanupSwitches,
    printSwitches,
) where

import Control.Monad (forM_, unless, when)
import System.Directory (doesPathExist)

-- | GPIO pin number
type Pin = Int

-- | Switch state
data SwitchState = SwitchState
    { pin :: Pin
    , isClosed :: Bool
    }
    deriving (Show, Eq)

-- | GPIO sysfs paths
gpioExport :: FilePath
gpioExport = "/sys/class/gpio/export"

gpioUnexport :: FilePath
gpioUnexport = "/sys/class/gpio/unexport"

gpioBase :: FilePath
gpioBase = "/sys/class/gpio"

-- | Default GPIO pins (P8_7 through P8_10 on BBB)
defaultPins :: [Pin]
defaultPins = [66, 67, 68, 69]

-- | Export a GPIO pin
exportPin :: Pin -> IO ()
exportPin p = do
    let path = gpioBase ++ "/gpio" ++ show p
    exists <- doesPathExist path
    unless exists $ writeFile gpioExport (show p)

-- | Unexport a GPIO pin
unexportPin :: Pin -> IO ()
unexportPin p = do
    let path = gpioBase ++ "/gpio" ++ show p
    exists <- doesPathExist path
    when exists $ writeFile gpioUnexport (show p)

-- | Set GPIO direction
setDirection :: Pin -> String -> IO ()
setDirection p dir = do
    let path = gpioBase ++ "/gpio" ++ show p ++ "/direction"
    writeFile path dir

-- | Set GPIO value
setValue :: Pin -> Int -> IO ()
setValue p val = do
    let path = gpioBase ++ "/gpio" ++ show p ++ "/value"
    writeFile path (show val)

-- | Initialize a single switch
initSwitch :: Pin -> IO SwitchState
initSwitch p = do
    exportPin p
    setDirection p "out"
    setValue p 0
    return $ SwitchState p False

-- | Initialize multiple switches
initSwitches :: [Pin] -> IO [SwitchState]
initSwitches = mapM initSwitch

-- | Close the switch (short the wires)
closeSwitch :: SwitchState -> IO SwitchState
closeSwitch s = do
    setValue (pin s) 1
    return $ s{isClosed = True}

-- | Open the switch (disconnect the wires)
openSwitch :: SwitchState -> IO SwitchState
openSwitch s = do
    setValue (pin s) 0
    return $ s{isClosed = False}

-- | Set switch state
setSwitch :: SwitchState -> Bool -> IO SwitchState
setSwitch s closed
    | closed = closeSwitch s
    | otherwise = openSwitch s

-- | Cleanup all switches
cleanupSwitches :: [SwitchState] -> IO ()
cleanupSwitches switches = forM_ switches $ \s -> do
    setValue (pin s) 0
    unexportPin (pin s)

-- | Print switch states
printSwitches :: [SwitchState] -> IO ()
printSwitches switches = forM_ (zip [0 :: Int ..] switches) $ \(i, s) -> do
    let state = if isClosed s then "CLOSED" else "OPEN"
    putStrLn $ "Switch " ++ show i ++ " (GPIO " ++ show (pin s) ++ "): " ++ state
